--- openssh-7.6p1/sshd.c	2018-01-18 12:15:14.623936099 -0800
+++ openssh-7.6p1-fips/sshd.c	2018-01-18 12:14:45.375305862 -0800
@@ -1348,6 +1348,20 @@
 #endif /* IP_OPTIONS */
 }
 
+static void set_fips_mode()
+{
+    error("set_fips_mode");
+    if (access("/tmp/.fipsModeOn", F_OK) != 0) return;
+
+    if (!FIPS_mode_set(1)) {
+        ERR_load_crypto_strings();
+        ERR_print_errors_fp(stderr);
+        error("failed in set_fips_mode in sshd.c");
+    } else {
+        error("succeeded in set_fips_mode in sshd.c");
+    }
+}
+
 /*
  * Main program for the daemon.
  */
@@ -1377,6 +1391,7 @@
 #ifdef HAVE_SECUREWARE
 	(void)set_auth_parameters(ac, av);
 #endif
+    set_fips_mode();
 	__progname = ssh_get_progname(av[0]);
 
 	/* Save argv. Duplicate so setproctitle emulation doesn't clobber it */
