diff --git a/networking/udhcp/dhcpc.c b/networking/udhcp/dhcpc.c
index 331f13a..8e8cf56 100644
--- a/networking/udhcp/dhcpc.c
+++ b/networking/udhcp/dhcpc.c
@@ -728,6 +728,17 @@ static NOINLINE int send_discover(uint32_t requested)
 	 */
 	init_packet(&packet, DHCPDISCOVER);
 
+        bb_info_msg("send_discover: xid %d", packet.xid);
+        char dhcpxid[256];
+        memset(dhcpxid,0,sizeof(dhcpxid));
+        snprintf(dhcpxid,sizeof(dhcpxid),"/tmp/dhcp.%s.xid",client_data.interface);
+        FILE* fp = fopen(dhcpxid,"w");
+        if(fp)
+        {
+                fprintf(fp,"%d",packet.xid);
+                fclose(fp);
+        }
+
 	if (requested)
 		udhcp_add_simple_option(&packet, DHCP_REQUESTED_IP, requested);
 
