--- busybox-1.22.1/sysklogd/syslogd.c	2017-10-06 14:04:33.655888755 -0700
+++ busybox-1.22.1-ltrx/sysklogd/syslogd.c	2017-10-07 00:23:05.097303519 -0700
@@ -1004,6 +1004,17 @@
 				rh->remoteFD = try_to_resolve_remote(rh);
 				if (rh->remoteFD == -1)
 					continue;
+                int local_port = 0;//any
+                char* port = getenv("SYSLOGD_LOCALPORT");
+                if(port)
+                {
+                    local_port = atoi(port);
+                    struct sockaddr_in local;
+                    local.sin_family = AF_INET;
+                    local.sin_port = htons(local_port);
+                    local.sin_addr.s_addr = htonl(INADDR_ANY);
+                    xbind(rh->remoteFD, (struct sockaddr *) &local, sizeof(local));
+                } 
 			}
             // LTRX: check severity code with configured log level
             {
