Index: openssl-1.0.2u/crypto/evp/e_des3.c
===================================================================
--- openssl-1.0.2u.orig/crypto/evp/e_des3.c
+++ openssl-1.0.2u/crypto/evp/e_des3.c
@@ -214,7 +214,7 @@ static int des_ede3_cfb1_cipher(EVP_CIPH
 
     if (!EVP_CIPHER_CTX_test_flags(ctx, EVP_CIPH_FLAG_LENGTH_BITS))
             inl *= 8;
-    for (n = 0; n < inl; ++n) {
+    for (n = 0; n * 8 < inl; ++n) {
         c[0] = (in[n / 8] & (1 << (7 - n % 8))) ? 0x80 : 0;
         DES_ede3_cfb_encrypt(c, d, 1, 1,
                              &data(ctx)->ks1, &data(ctx)->ks2,
