--- openssl-1.0.2u/crypto/ec/ec_asn1.c
+++ openssl-1.0.2u/crypto/ec/ec_asn1.c
@@ -867,7 +867,10 @@ static EC_GROUP *ec_asn1_parameters2group(const ECPARAMETERS *params)
         ret->seed_len = params->curve->seed->length;
     }
 
-    if (!params->order || !params->base || !params->base->data) {
+    if (!params->order
+            || !params->base
+            || !params->base->data
+            || !params->base->length) {
         ECerr(EC_F_EC_ASN1_PARAMETERS2GROUP, EC_R_ASN1_ERROR);
         goto err;
     }
