--- libwebsockets-v2.4.2/lib/client/client-handshake.c	2018-05-15 10:49:56.988198619 -0700
+++ libwebsockets-v2.4.2-ltrx/lib/client/client-handshake.c	2018-05-15 10:49:16.991428411 -0700
@@ -267,11 +267,18 @@
 
 		lws_set_timeout(wsi, PENDING_TIMEOUT_AWAITING_CONNECT_RESPONSE,
 				AWAITING_TIMEOUT);
+        
+		int local_port = 0;//any
+		char* lws_port = getenv("LIBWEBSOCKETS_LOCALPORT");
+		if(lws_port)
+		{
+			local_port = atoi(lws_port);
+		}
 
 		iface = lws_hdr_simple_ptr(wsi, _WSI_TOKEN_CLIENT_IFACE);
 
 		if (iface) {
-			n = lws_socket_bind(wsi->vhost, wsi->desc.sockfd, 0, iface);
+			n = lws_socket_bind(wsi->vhost, wsi->desc.sockfd, local_port, iface);
 			if (n < 0) {
 				cce = "unable to bind socket";
 				goto failed;
