--- libwebsockets-v2.4.2/lib/client/client-handshake.c	2018-05-23 17:22:42.076371546 -0700
+++ libwebsockets-v2.4.2-ltrx/lib/client/client-handshake.c	2018-05-23 17:26:58.449474655 -0700
@@ -267,7 +267,7 @@
 
 		lws_set_timeout(wsi, PENDING_TIMEOUT_AWAITING_CONNECT_RESPONSE,
 				AWAITING_TIMEOUT);
-        
+
 		int local_port = 0;//any
 		char* lws_port = getenv("LIBWEBSOCKETS_LOCALPORT");
 		if(lws_port)
@@ -277,7 +277,8 @@
 
 		iface = lws_hdr_simple_ptr(wsi, _WSI_TOKEN_CLIENT_IFACE);
 
-		if (iface) {
+		//LTRX: iface can be NULL but we need lws_socket_bind to bind to a local port
+		if (iface || local_port) {
 			n = lws_socket_bind(wsi->vhost, wsi->desc.sockfd, local_port, iface);
 			if (n < 0) {
 				cce = "unable to bind socket";
@@ -1048,7 +1049,7 @@
 		/* the length of the address, excluding port */
 		pt->serv_buf[n] = strlen(wsi->u.hdr.stash->address);
 		break;
-		
+
 	default:
 		return;
 	}
