--- libwebsockets-v2.4.2/lib/libwebsockets.c	2018-05-17 16:14:16.924809634 -0700
+++ libwebsockets-v2.4.2-ltrx/lib/libwebsockets.c	2018-05-17 16:19:56.387268995 -0700
@@ -1480,8 +1480,9 @@
 	/* we have to deal with a possible redundant leading http:// */
 	if (!strncmp(proxy, "http://", 7))
 		proxy += 7;
-
-	p = strchr(proxy, '@');
+    //proxy: allow use of at char in passwords
+    //https://github.com/warmcat/libwebsockets/commit/a55a16e5a6f7d616545fbeaab10e063fb212899e
+	p = strrchr(proxy, '@');
 	if (p) { /* auth is around */
 
 		if ((unsigned int)(p - proxy) > sizeof(authstring) - 1)
@@ -1543,7 +1544,9 @@
 	vhost->socks_user[0] = '\0';
 	vhost->socks_password[0] = '\0';
 
-	p_at = strchr(socks, '@');
+    //proxy: allow use of at char in passwords
+    //https://github.com/warmcat/libwebsockets/commit/a55a16e5a6f7d616545fbeaab10e063fb212899e
+	p_at = strrchr(socks, '@');
 	if (p_at) { /* auth is around */
 		if ((unsigned int)(p_at - socks) > (sizeof(user)
 			+ sizeof(password) - 2)) {
