--- mrouted-3.9.6/route.c	2017-06-09 16:33:52.558396034 -0700
+++ mrouted-3.9.6-ltrx/route.c	2017-06-09 16:35:27.089919611 -0700
@@ -1251,8 +1251,10 @@
     int metric;
 
     p = send_buf + MIN_IP_HEADER_LEN + IGMP_MINLEN;
-
-    for (r = start_rt; r != routing_table; r = r->rt_prev) {
+    /*
+    0001-DVMRP-report-missing-subnet.patch
+    */
+    for (r = start_rt; r != NULL; r = r->rt_prev) {
 	if (which_routes == CHANGED_ROUTES && !(r->rt_flags & RTF_CHANGED)) {
 	    nrt++;
 	    continue;
